tests/semantic_analyser.cpp

```diff
@@ -439,10 +440,37 @@ BEGIN { @x[1] = 0; @x[2,3]; }
       @x["b", 2, kstack];
     })",
              R"(
-stdin:3:7-25: ERROR: Argument mismatch for @x: trying to access with arguments: [string[2], unsigned int64, kstack] when map expects arguments: [unsigned int64, string[2], kstack]
+stdin:3:7-25: ERROR: Argument mismatch for @x: trying to access with arguments: '(string[2],unsigned int64,kstack)' when map expects arguments: '(unsigned int64,string[2],kstack)'
       @x["b", 2, kstack];
       ~~~~~~~~~~~~~~~~~~
 )");
+
+  test("BEGIN { @map[1, 2] = 1; for ($kv : @map) { @map[$kv.0] = 2; } }");
+
+  test_error(
+      R"(BEGIN { @map[1, 2] = 1; for ($kv : @map) { @map[$kv.0.0] = 2; } })",
+      R"(
+stdin:1:45-58: ERROR: Argument mismatch for @map: trying to access with arguments: 'unsigned int64' when map expects arguments: '(unsigned int64,unsigned int64)'
+BEGIN { @map[1, 2] = 1; for ($kv : @map) { @map[$kv.0.0] = 2; } }
+                                            ~~~~~~~~~~~~~
+)");
+
+  test(R"(BEGIN { $a = (3, "hi"); @map[1, "by"] = 1; @map[$a] = 2; })");
+
+  test_error(R"(BEGIN { $a = (3, "hello"); @map[1, "by"] = 1; @map[$a] = 2; })",
+             R"(
+stdin:1:47-55: ERROR: Argument mismatch for @map: trying to access with arguments: '(int64,string[6])' when map expects arguments: '(unsigned int64,string[3])'
+BEGIN { $a = (3, "hello"); @map[1, "by"] = 1; @map[$a] = 2; }
+                                              ~~~~~~~~
+)");
+
+  test_error(
+      R"(BEGIN { $a = (3, "hello"); $b = (4, "hi"); @map[$a] = 1; @map[$b] = 2; })",
+      R"(
+stdin:1:58-66: ERROR: Argument mismatch for @map: trying to access with arguments: '(int64,string[3])' when map expects arguments: '(int64,string[6])'
+BEGIN { $a = (3, "hello"); $b = (4, "hi"); @map[$a] = 1; @map[$b] = 2; }
+                                                         ~~~~~~~~
+)");
```

Comment:
These expected failures are a bit annoying - the sizes of strings doesn't make a difference anywhere else so it's a bit strange to require them to exactly match here